Add pipe actions

NathanReb opened the pull request:
Fixes #428 

This PR introduces `pipe-outputs`, `pipe-stdout` and `pipe-stderr` to the action language.

The tests don't cover much atm. In particular I'd like to make sure I got the `Action_to_sh` part right but I'm unsure what's the best way to test it.

I also believe I need to fix the stdout/stderr-only cases by adding a `multi_use` to the one that's not being piped.

NathanReb committed:
Add pipe actions to the language

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

diml commented:
@NathanReb is this ready for review? If not, could you add `[WIP]` in the title?

NathanReb commented:
To be honest I'm not even sure myself, I'll try to add the `multi_use` but then it should be ready for review!

NathanReb committed:
Share non-piped outputs channel in pipe actions

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

NathanReb committed:
Add a test and fix pipe actions conversion to shell

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

NathanReb commented:
I added the proper `multi_use` channels and also added a test for the shell conversion.

If you look at that test you'll note that the shell conversion of echo doesn't seem to work very well. Also it seems that the shell conversions swallow quotes when they shouldn't.

I'll propabaly change that test so that it doesn't depend on `echo` and `run` to be properly converted but we should look into fixing those!

NathanReb committed:
Simplify pipe-actions shell conversion test case

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

NathanReb commented:
I just pushed a commit to simplify the shell conversion test so if you want to see the bugs I mentioned you should take a look at [this](https://github.com/ocaml/dune/pull/3392/commits/00030c7982247ef4a8a6c1c3f903230fc960fbd0#diff-f4995aeb316f6aef9c4922ae4ac50c49R49).

NathanReb commented:
Ok it seems like it didn't swallow single quotes when translating `(run grep "'a\\|b'")`. I just did not promote the last change.

The bug with echo is real though!

jeremiedimino commented on src/dune/action_exec.ml:
```diff
@@ -387,6 +399,40 @@ and exec_list ts ~ectx ~eenv =
+  in
+  let multi_use_eenv =
+    match outputs with
+    | Outputs -> eenv
```
That looks wrong

jeremiedimino replied:
Never-mind, it is correct

NathanReb replied:
Do you want me to add a comment there to clarify why it's done like this?

jeremiedimino replied:
No that's fine, I just read it too fast.

jeremiedimino commented on src/dune/action_exec.ml:
```diff
@@ -387,6 +399,40 @@ and exec_list ts ~ectx ~eenv =
+and exec_pipe outputs ts ~ectx ~eenv =
+  let tmp_file () =
+    Path.of_string
+      (Filename.temp_file "pipe-action-"
```
In `process.ml`, there is a `Temp` sub-module that ensures that temporary files are deleted in case of abnormal exit. We should extract it to reuse it here.

jeremiedimino committed:
Tweak the pipe test

Use tr instead of grep:
- avoid some nasty quoting
- it's easier to observe the effect of transformations

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>
Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

jeremiedimino commented on src/dune/action_exec.ml:
```diff
@@ -387,6 +399,40 @@ and exec_list ts ~ectx ~eenv =
+    Path.of_string
+      (Filename.temp_file "pipe-action-"
+         ("." ^ Action.Outputs.to_string outputs))
+  in
```
Right now, nothing is deleting the temporary file BTW. And indeed, they are left in `/tmp` after dune exits.

jeremiedimino commented:
This makes me think that we should improvement our management of temporary files: #3432

NathanReb commented:
I just rebased and fixed the conflicts!

I'm guessing I should fix the temp file mgmt issue first in a separate PR or is it okay to do it here?

jeremiedimino commented:
A separate PR is better

NathanReb commented:
I just rebased on top of #3521 and slightly improved the tmp file management for pipe actions by deleting intermediate files as it goes.

NathanReb committed:
Document the new pipe actions

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

NathanReb committed:
Bump language to 2.7

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>
